File: lib/game/threat.ts
/**
 * Threat director (pure rules) — Roadmap 4, Military.
 *
 * Raid pressure builds as the colony gets richer, larger, better-defended and
 * older, then discharges as a raid when it crosses a threshold. Young colonies
 * get a grace window so a fresh run isn't sacked in its first minutes. Enemy
 * warbands scale to what they're attacking — wealth and the standing warrior
 * count both make the next raid bigger — so success invites a stiffer test.
 *
 * All randomness is injected (a `roll` in [0,1)) so the caller drives it from
 * the seeded chain and raids stay deterministic under setTestRngSeed.
 */

export interface ThreatSnapshot {
	/** Total stored value the colony presents as loot (see {@link colonyWealth}). */
	wealth: number;
	/** Living cats. */
	population: number;
	/** Trained warriors currently standing. */
	warriors: number;
	/** Game-seconds since the run started (accelerated clock). */
	colonyAgeSec: number;
}

/** Grace window: no pressure builds for this many game-seconds into a run. */
export const RAID_GRACE_SEC = 8 * 3600;
/** Pressure at which a raid launches. */
export const RAID_SPAWN_THRESHOLD = 100;
/** Largest warband the director will field at once. */
export const MAX_RAID_SIZE = 12;
/** Base strength of a single raider before scaling. */
export const RAIDER_BASE_STRENGTH = 30;
/**
 * A colony poorer than this presents no extra raiders beyond the lone scout —
 * a fresh settlement (~240 stored value) sits below it, so its first raids are
 * a single weak raider a militia turns away without breaking stride.
 */
export const RAID_WEALTH_FLOOR = 250;
/** Population above this baseline (the starter roster) adds raiders. */
export const RAID_POP_FLOOR = 20;

/** Threat-indicator band for the HUD. */
export type ThreatBand = "calm" | "rising" | "imminent";

/**
 * Loot value a colony presents — the sum the threat curve reads as "wealth".
 * Refined goods and forged gear are worth more than raw stores.
 */
export function colonyWealth(resources: {
	food?: number;
	water?: number;
	herbs?: number;
	materials?: number;
	refined?: number;
	weapons?: number;
	armor?: number;
}): number {
	return (
		(resources.food ?? 0) +
		(resources.water ?? 0) +
		(resources.herbs ?? 0) +
		(resources.materials ?? 0) +
		(resources.refined ?? 0) * 3 +
		(resources.weapons ?? 0) * 5 +
		(resources.armor ?? 0) * 5
	);
}

/**
 * Pressure gained per game-hour. Zero during the grace window; otherwise a
 * baseline plus contributions from wealth (sub-linear so a hoard doesn't
 * explode), population, the standing warrior count (enemies scale to your
 * defenses), and colony age (a slow, capped creep).
 */
export function threatRatePerHour(s: ThreatSnapshot): number {
	if (s.colonyAgeSec < RAID_GRACE_SEC) {
		return 0;
	}
	const wealthTerm = Math.sqrt(Math.max(0, s.wealth)) * 0.12;
	const popTerm = Math.max(0, s.population) * 0.25;
	const warriorTerm = Math.max(0, s.warriors) * 0.5;
	const ageTerm = Math.min(10, (s.colonyAgeSec / 3600) * 0.04);
	return 1 + wealthTerm + popTerm + warriorTerm + ageTerm;
}

/** Add the pressure earned over `elapsedGameSec` to the running total. */
export function accrueThreat(
	pressure: number,
	s: ThreatSnapshot,
	elapsedGameSec: number,
): number {
	if (elapsedGameSec <= 0) {
		return Math.max(0, pressure);
	}
	const gained = threatRatePerHour(s) * (elapsedGameSec / 3600);
	return Math.max(0, pressure + gained);
}

/** A raid launches once accrued pressure reaches the threshold. */
export function shouldSpawnRaid(pressure: number): boolean {
	return pressure >= RAID_SPAWN_THRESHOLD;
}

/**
 * HUD threat band from the current pressure: calm below a third of the
 * threshold, rising up to two thirds, imminent above.
 */
export function threatBand(pressure: number): ThreatBand {
	if (pressure >= RAID_SPAWN_THRESHOLD * (2 / 3)) {
		return "imminent";
	}
	if (pressure >= RAID_SPAWN_THRESHOLD / 3) {
		return "rising";
	}
	return "calm";
}

export interface RaidPlan {
	/** Number of raider units in the warband. */
	count: number;
	/** Strength (and starting hp) of each raider. */
	strengthEach: number;
}

/**
 * Size the warband for a snapshot. A raid always fields at least a lone scout;
 * extra raiders come only as the colony matures — from its standing warriors,
 * the wealth it holds above a starter floor, and the population it has grown
 * past the starter roster (all clamped to {@link MAX_RAID_SIZE}). A fresh
 * settlement — no warriors, starter stores, starter roster — therefore faces a
 * single raider, and each raider's strength creeps up with colony age so
 * late-game raids bite harder.
 */
export function planRaid(s: ThreatSnapshot): RaidPlan {
	const fromWarriors = Math.floor(Math.max(0, s.warriors) * 0.6);
	const fromWealth = Math.floor(
		Math.sqrt(Math.max(0, s.wealth - RAID_WEALTH_FLOOR)) / 10,
	);
	const fromPop = Math.floor(Math.max(0, s.population - RAID_POP_FLOOR) / 15);
	const count = Math.max(
		1,
		Math.min(MAX_RAID_SIZE, 1 + fromWarriors + fromWealth + fromPop),
	);
	const ageBonus = Math.min(40, (s.colonyAgeSec / 3600) * 0.5);
	const strengthEach = Math.round(RAIDER_BASE_STRENGTH + ageBonus);
	return { count, strengthEach };
}

export interface RaidOutcome {
	defendersWin: boolean;
	/** 0..1 share of stores the raiders carry off on a successful raid. */
	lootFraction: number;
	/** Cats killed defending (only on a loss, and only when it's close). */
	defenderCasualties: number;
	/** How decisively it went, for flavor: >1 rout, ~1 close, <1 defeat. */
	margin: number;
}

/** Below this power ratio a defeat also costs a defender's life. */
const CASUALTY_RATIO = 0.6;
/**
 * Hard ceiling on cats a single lost raid can kill. One bad fight must never
 * cascade into a wipe: a colony can lose several raids and still live, so
 * collapse comes from neglect (starvation, thirst), not from being outnumbered
 * once. See {@link RaidOutcome.defenderCasualties}.
 */
export const MAX_RAID_CASUALTIES = 1;
/**
 * Most of the stores a single raid can carry off. Bounds the theft alongside
 * {@link MAX_RAID_CASUALTIES} so a sacked colony keeps the majority of every
 * store and can recover.
 */
const MAX_LOOT_FRACTION = 0.3;

/**
 * Resolve a raid: mustered defense power vs the raiders' remaining strength.
 * A roll in [0,1) adds +/-25% swing so evenly-matched fights aren't fully
 * predetermined. Defenders win when their swung power meets or beats the
 * raiders; a rout leaves no casualties, a defeat bleeds stores (and, if badly
 * outmatched, a life).
 */
export function resolveRaid(
	defensePower: number,
	raiderPower: number,
	roll: number,
): RaidOutcome {
	const swing = 0.75 + 0.5 * Math.min(0.999999, Math.max(0, roll));
	const effective = defensePower * swing;
	const enemy = Math.max(1, raiderPower);
	const margin = effective / enemy;
	const defendersWin = effective >= enemy;

	if (defendersWin) {
		return {
			defendersWin: true,
			lootFraction: 0,
			defenderCasualties: 0,
			margin,
		};
	}

	// The worse the defenders lose, the more the raiders haul off.
	const shortfall = 1 - Math.min(1, margin);
	const lootFraction = Math.min(MAX_LOOT_FRACTION, 0.1 + shortfall * 0.3);
	const defenderCasualties = margin < CASUALTY_RATIO ? MAX_RAID_CASUALTIES : 0;
	return { defendersWin: false, lootFraction, defenderCasualties, margin };
}

File: crates/cat-sim/src/threat.rs
//! Raid threat director ported from `lib/game/threat.ts`.

use serde::{Deserialize, Serialize};

use crate::entities::Resources;

/// Grace window: no pressure builds for this many game-seconds into a run.
pub const RAID_GRACE_SEC: f64 = 8.0 * 3600.0;
/// Pressure at which a raid launches.
pub const RAID_SPAWN_THRESHOLD: f64 = 100.0;
/// Largest warband the director will field at once.
pub const MAX_RAID_SIZE: f64 = 12.0;
/// Base strength of a single raider before scaling.
pub const RAIDER_BASE_STRENGTH: f64 = 30.0;
/// Wealth floor before stored value adds extra raiders.
pub const RAID_WEALTH_FLOOR: f64 = 250.0;
/// Population above this baseline adds extra raiders.
pub const RAID_POP_FLOOR: f64 = 20.0;
/// Hard ceiling on cats a single lost raid can kill.
pub const MAX_RAID_CASUALTIES: u32 = 1;

const CASUALTY_RATIO: f64 = 0.6;
const MAX_LOOT_FRACTION: f64 = 0.3;

/// Total stored value the colony presents as loot.
pub trait ThreatStores {
    fn food(&self) -> f64;
    fn water(&self) -> f64;
    fn herbs(&self) -> f64;
    fn materials(&self) -> f64;
    fn refined(&self) -> f64;
    fn weapons(&self) -> f64;
    fn armor(&self) -> f64;
}

impl<T> ThreatStores for &T
where
    T: ThreatStores + ?Sized,
{
    fn food(&self) -> f64 {
        (*self).food()
    }

    fn water(&self) -> f64 {
        (*self).water()
    }

    fn herbs(&self) -> f64 {
        (*self).herbs()
    }

    fn materials(&self) -> f64 {
        (*self).materials()
    }

    fn refined(&self) -> f64 {
        (*self).refined()
    }

    fn weapons(&self) -> f64 {
        (*self).weapons()
    }

    fn armor(&self) -> f64 {
        (*self).armor()
    }
}

impl ThreatStores for Resources {
    fn food(&self) -> f64 {
        self.food + self.fish
    }

    fn water(&self) -> f64 {
        self.water
    }

    fn herbs(&self) -> f64 {
        self.herbs
    }

    fn materials(&self) -> f64 {
        self.materials
    }

    fn refined(&self) -> f64 {
        self.refined
    }

    fn weapons(&self) -> f64 {
        self.weapons
    }

    fn armor(&self) -> f64 {
        self.armor
    }
}

/// Minimal resource shape accepted by [`colony_wealth`].
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize, Default)]
pub struct ThreatResources {
    pub food: f64,
    pub water: f64,
    pub herbs: f64,
    pub materials: f64,
    pub refined: f64,
    pub weapons: f64,
    pub armor: f64,
}

impl ThreatStores for ThreatResources {
    fn food(&self) -> f64 {
        self.food
    }

    fn water(&self) -> f64 {
        self.water
    }

    fn herbs(&self) -> f64 {
        self.herbs
    }

    fn materials(&self) -> f64 {
        self.materials
    }

    fn refined(&self) -> f64 {
        self.refined
    }

    fn weapons(&self) -> f64 {
        self.weapons
    }

    fn armor(&self) -> f64 {
        self.armor
    }
}

/// Threat inputs for the raid director.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ThreatSnapshot {
    pub wealth: f64,
    pub population: f64,
    pub warriors: f64,
    pub colony_age_sec: f64,
}

/// Threat-indicator band for the HUD.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ThreatBand {
    Calm,
    Rising,
    Imminent,
}

impl ThreatBand {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Calm => "calm",
            Self::Rising => "rising",
            Self::Imminent => "imminent",
        }
    }
}

/// Number and strength of raider units in a planned warband.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RaidPlan {
    pub count: f64,
    pub strength_each: f64,
}

/// Deterministic combat result for a raid.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RaidOutcome {
    pub defenders_win: bool,
    pub loot_fraction: f64,
    pub defender_casualties: u32,
    pub margin: f64,
}

/// Loot value a colony presents to the threat curve.
#[must_use]
pub fn colony_wealth(resources: impl ThreatStores) -> f64 {
    resources.food()
        + resources.water()
        + resources.herbs()
        + resources.materials()
        + resources.refined() * 3.0
        + resources.weapons() * 5.0
        + resources.armor() * 5.0
}

/// Pressure gained per game-hour.
#[must_use]
pub fn threat_rate_per_hour(snapshot: ThreatSnapshot) -> f64 {
    if snapshot.colony_age_sec < RAID_GRACE_SEC {
        return 0.0;
    }

    let wealth_term = js_max(0.0, snapshot.wealth).sqrt() * 0.12;
    let pop_term = js_max(0.0, snapshot.population) * 0.25;
    let warrior_term = js_max(0.0, snapshot.warriors) * 0.5;
    let age_term = js_min(10.0, (snapshot.colony_age_sec / 3600.0) * 0.04);

    1.0 + wealth_term + pop_term + warrior_term + age_term
}

/// Add the pressure earned over `elapsed_game_sec` to the running total.
#[must_use]
pub fn accrue_threat(pressure: f64, snapshot: ThreatSnapshot, elapsed_game_sec: f64) -> f64 {
    if elapsed_game_sec <= 0.0 {
        return js_max(0.0, pressure);
    }

    let gained = threat_rate_per_hour(snapshot) * (elapsed_game_sec / 3600.0);
    js_max(0.0, pressure + gained)
}

/// A raid launches once accrued pressure reaches the threshold.
#[must_use]
pub fn should_spawn_raid(pressure: f64) -> bool {
    pressure >= RAID_SPAWN_THRESHOLD
}

/// HUD threat band from the current pressure.
#[must_use]
pub fn threat_band(pressure: f64) -> ThreatBand {
    if pressure >= RAID_SPAWN_THRESHOLD * (2.0 / 3.0) {
        return ThreatBand::Imminent;
    }
    if pressure >= RAID_SPAWN_THRESHOLD / 3.0 {
        return ThreatBand::Rising;
    }
    ThreatBand::Calm
}

/// Size the warband for a snapshot.
#[must_use]
pub fn plan_raid(snapshot: ThreatSnapshot) -> RaidPlan {
    let from_warriors = (js_max(0.0, snapshot.warriors) * 0.6).floor();
    let from_wealth = (js_max(0.0, snapshot.wealth - RAID_WEALTH_FLOOR).sqrt() / 10.0).floor();
    let from_pop = (js_max(0.0, snapshot.population - RAID_POP_FLOOR) / 15.0).floor();
    let count = js_max(
        1.0,
        js_min(MAX_RAID_SIZE, 1.0 + from_warriors + from_wealth + from_pop),
    );
    let age_bonus = js_min(40.0, (snapshot.colony_age_sec / 3600.0) * 0.5);
    let strength_each = js_round(RAIDER_BASE_STRENGTH + age_bonus);

    RaidPlan {
        count,
        strength_each,
    }
}

/// Resolve a raid using injected randomness.
#[must_use]
pub fn resolve_raid(defense_power: f64, raider_power: f64, roll: f64) -> RaidOutcome {
    let swing = 0.75 + 0.5 * js_min(0.999_999, js_max(0.0, roll));
    let effective = defense_power * swing;
    let enemy = js_max(1.0, raider_power);
    let margin = effective / enemy;
    let defenders_win = effective >= enemy;

    if defenders_win {
        return RaidOutcome {
            defenders_win: true,
            loot_fraction: 0.0,
            defender_casualties: 0,
            margin,
        };
    }

    let shortfall = 1.0 - js_min(1.0, margin);
    let loot_fraction = js_min(MAX_LOOT_FRACTION, 0.1 + shortfall * 0.3);
    let defender_casualties = if margin < CASUALTY_RATIO {
        MAX_RAID_CASUALTIES
    } else {
        0
    };

    RaidOutcome {
        defenders_win: false,
        loot_fraction,
        defender_casualties,
        margin,
    }
}

fn js_max(left: f64, right: f64) -> f64 {
    if left.is_nan() || right.is_nan() {
        f64::NAN
    } else if left >= right {
        left
    } else {
        right
    }
}

fn js_min(left: f64, right: f64) -> f64 {
    if left.is_nan() || right.is_nan() {
        f64::NAN
    } else if left <= right {
        left
    } else {
        right
    }
}

fn js_round(value: f64) -> f64 {
    (value + 0.5).floor()
}

#[cfg(test)]
mod tests {
    use crate::entities::Resources;

    use super::{
        MAX_RAID_CASUALTIES, MAX_RAID_SIZE, RAID_GRACE_SEC, RAID_SPAWN_THRESHOLD, RaidOutcome,
        RaidPlan, ThreatBand, ThreatResources, ThreatSnapshot, accrue_threat, colony_wealth,
        plan_raid, resolve_raid, should_spawn_raid, threat_band, threat_rate_per_hour,
    };

    fn snap() -> ThreatSnapshot {
        ThreatSnapshot {
            wealth: 500.0,
            population: 20.0,
            warriors: 2.0,
            colony_age_sec: RAID_GRACE_SEC + 3600.0,
        }
    }

    fn assert_f64_close(actual: f64, expected: f64, label: &str) {
        assert!(
            (actual - expected).abs() <= 1e-12,
            "{label}: actual {actual:?}, expected {expected:?}",
        );
    }

    #[test]
    fn colony_wealth_weights_refined_and_gear() {
        let resources = ThreatResources {
            food: 10.0,
            water: 20.0,
            herbs: 30.0,
            materials: 40.0,
            refined: 5.0,
            weapons: 2.0,
            armor: 3.0,
        };

        assert_eq!(colony_wealth(resources), 140.0);
        assert_eq!(colony_wealth(ThreatResources::default()), 0.0);
        assert_eq!(
            colony_wealth(Resources {
                food: 10.0,
                fish: 0.0,
                water: 20.0,
                herbs: 30.0,
                catnip: 0.0,
                grain: 0.0,
                flour: 0.0,
                preserves: 0.0,
                medicine: 0.0,
                brew: 0.0,
                materials: 40.0,
                stone: 0.0,
                refined: 5.0,
                weapons: 2.0,
                armor: 3.0,
                planks: 0.0,
                logs: 0.0,
                lumber: 0.0,
                blocks: 0.0,
                tools: 0.0,
                fibre: 0.0,
                hide: 0.0,
                bone: 0.0,
                cloth: 0.0,
                leather: 0.0,
                ore: 0.0,
                gem: 0.0,
                clay: 0.0,
                sand: 0.0,
                metal: 0.0,
                blessings: 999.0,
            }),
            140.0
        );
    }

    #[test]
    fn threat_rate_obeys_grace_and_terms() {
        assert_eq!(
            threat_rate_per_hour(ThreatSnapshot {
                colony_age_sec: 0.0,
                ..snap()
            }),
            0.0
        );
        assert_eq!(
            threat_rate_per_hour(ThreatSnapshot {
                colony_age_sec: RAID_GRACE_SEC - 1.0,
                ..snap()
            }),
            0.0
        );

        assert_f64_close(
            threat_rate_per_hour(ThreatSnapshot {
                colony_age_sec: RAID_GRACE_SEC,
                ..snap()
            }),
            10.003_281_572_999_748,
            "grace boundary rate",
        );
        assert_f64_close(
            threat_rate_per_hour(snap()),
            10.043_281_572_999_747,
            "default rate",
        );
    }

    #[test]
    fn accrue_threat_adds_elapsed_game_time_and_clamps_nonpositive_ticks() {
        assert_f64_close(
            accrue_threat(0.0, snap(), 3600.0),
            10.043_281_572_999_747,
            "one hour gained pressure",
        );
        assert_eq!(
            accrue_threat(
                10.0,
                ThreatSnapshot {
                    colony_age_sec: 0.0,
                    ..snap()
                },
                3600.0
            ),
            10.0
        );
        assert_eq!(accrue_threat(-5.0, snap(), 0.0), 0.0);
    }

    #[test]
    fn threshold_and_bands_match_thirds() {
        assert!(!should_spawn_raid(RAID_SPAWN_THRESHOLD - 0.01));
        assert!(should_spawn_raid(RAID_SPAWN_THRESHOLD));

        assert_eq!(threat_band(0.0), ThreatBand::Calm);
        assert_eq!(threat_band(RAID_SPAWN_THRESHOLD / 3.0), ThreatBand::Rising);
        assert_eq!(
            threat_band(RAID_SPAWN_THRESHOLD * (2.0 / 3.0)),
            ThreatBand::Imminent
        );
        assert_eq!(ThreatBand::Imminent.as_str(), "imminent");
    }

    #[test]
    fn plan_raid_matches_hand_derived_vectors() {
        let starter = plan_raid(ThreatSnapshot {
            wealth: 240.0,
            population: 20.0,
            warriors: 0.0,
            colony_age_sec: RAID_GRACE_SEC,
        });
        assert_eq!(
            starter,
            RaidPlan {
                count: 1.0,
                strength_each: 34.0,
            }
        );

        let default = plan_raid(snap());
        assert_eq!(
            default,
            RaidPlan {
                count: 3.0,
                strength_each: 35.0,
            }
        );

        let wealthy = plan_raid(ThreatSnapshot {
            wealth: 6000.0,
            population: 45.0,
            warriors: 8.0,
            colony_age_sec: RAID_GRACE_SEC + 30.0 * 3600.0,
        });
        assert_eq!(
            wealthy,
            RaidPlan {
                count: MAX_RAID_SIZE,
                strength_each: 49.0,
            }
        );
    }

    #[test]
    fn resolve_raid_matches_hand_derived_win_loss_and_caps() {
        assert_eq!(
            resolve_raid(1000.0, 100.0, 0.5),
            RaidOutcome {
                defenders_win: true,
                loot_fraction: 0.0,
                defender_casualties: 0,
                margin: 10.0,
            }
        );

        let loss = resolve_raid(50.0, 1000.0, 0.5);
        assert!(!loss.defenders_win);
        assert_eq!(loss.defender_casualties, MAX_RAID_CASUALTIES);
        assert_f64_close(loss.margin, 0.05, "loss margin");
        assert_f64_close(loss.loot_fraction, 0.3, "loss loot cap");

        let close = resolve_raid(100.0, 100.0, 0.0);
        assert!(!close.defenders_win);
        assert_eq!(close.defender_casualties, 0);
        assert_f64_close(close.margin, 0.75, "close margin");
        assert_f64_close(close.loot_fraction, 0.175, "close loot");

        assert!(resolve_raid(100.0, 100.0, 0.999).defenders_win);
    }

    #[test]
    fn nan_inputs_follow_javascript_min_max_comparisons() {
        assert!(
            threat_rate_per_hour(ThreatSnapshot {
                wealth: f64::NAN,
                ..snap()
            })
            .is_nan()
        );
        assert_eq!(threat_band(f64::NAN), ThreatBand::Calm);
        assert!(
            plan_raid(ThreatSnapshot {
                warriors: f64::NAN,
                ..snap()
            })
            .count
            .is_nan()
        );

        let outcome = resolve_raid(10.0, 100.0, f64::NAN);
        assert!(!outcome.defenders_win);
        assert!(outcome.loot_fraction.is_nan());
        assert_eq!(outcome.defender_casualties, 0);
        assert!(outcome.margin.is_nan());
    }
}
